cloud-ctrl-mgr: enable secure port 10258

sttts opened the pull request:
This PR enables authn+authz (delegated to the kube-apiserver) and the secure port 10258 for the cloud-controller-manager. In addition, the insecure port is disabled.

This is the counterpart PR to https://github.com/kubernetes/kubernetes/pull/64149.

Moreover, it adds integration test coverage for the `--port` and `--secure-port` flags, plus the testserver infrastructure to tests flags in general inside integration tests.

```release-note
Enable secure serving on port 10258 to cloud-controller-manager (configurable via `--secure-port`). Delegated authentication and authorization are to be configured using the same flags as for aggregated API servers. Without configuration, the secure port will only allow access to `/healthz`.
```

sttts committed:
cloud-controller-manager: enable delegated authz/authn if secure port is enabled

sttts committed:
cloud-controller-manager: enable secure ports 10258, deprecate insecure port

yue9944882 commented on cmd/cloud-controller-manager/app/options/options.go:
```diff
@@ -176,11 +173,13 @@ func (o *CloudControllerManagerOptions) ApplyTo(c *cloudcontrollerconfig.Config,
-	}
-	if err = o.Authorization.ApplyTo(&c.Authorization); err != nil {
-		return err
+	if o.SecureServing.BindPort != 0 || o.SecureServing.Listener != nil {
```
I think `o.SecureServing.BindPort <= 0` would be better, in case that it's a negative value. 😃

sttts replied:
validation checks that

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/options/deprecated_insecure_serving.go:
```diff
@@ -0,0 +1,164 @@
+
+	// ListenFunc can be overridden to create a custom listener, e.g. for mocking in tests.
+	// It defaults to options.CreateListener.
+	ListenFunc func(network, addr string) (net.Listener, int, error)
```
Seems like `ListenFunc` is only used for test. What if we remove this field?

sttts replied:
We added it to have a natural bootstrapping way with minimal difference to the real binary. We don't want to tweak the config.

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/options/deprecated_insecure_serving.go:
```diff
@@ -0,0 +1,164 @@
+
+	errors := []error{}
+
+	if s.BindPort < 0 || s.BindPort > 65335 {
```
[https://github.com/kubernetes/kubernetes/pull/67069/files#r208195357](https://github.com/kubernetes/kubernetes/pull/67069/files#r208195357)

just like what we do here `s.BindPort < 0`

yue9944882 commented on staging/src/k8s.io/apiserver/pkg/server/options/deprecated_insecure_serving.go:
```diff
@@ -0,0 +1,164 @@
+		return nil
+	}
+
+	secureLoopbackClientConfig, err := (*insecureServingInfo).NewLoopbackClientConfig()
```
we needn't indirect ptr of `insecureServingInfo ` here

sttts replied:
hmm, then I wonder whether this code is even correct (the if above).

dims commented:
/uncc

sttts commented:
/assign @andrewsykim

andrewsykim commented:
/lgtm for `cmd/cloud-controller-manager`

nikhita commented:
/remove-area custom-resources

sttts committed:
cloud-controller-manager: enable secure loopback

sttts committed:
cloud-controller-manager: add test server

sttts committed:
controller-managers: generalize authn/z test to cloud-controller-manager

luxas commented:
@sttts any update here?

sttts commented:
/assign @simo5 @enxebre 

Can you review from auth and cloud team point of view?

sttts commented:
/retest

simo5 commented:
@enj PTAL

sttts commented:
@enj note, this is based on https://github.com/kubernetes/kubernetes/pull/64149. Only the cloud-ctrl-manager commits are important. The actual wiring of https+authn/z is rather trivial. Most work is the integration tests, generalizing the one from kube-controller-manager.

sttts committed:
cloud-controller-manager: disable authn/z on insecure port

This is the old behaviour and we did not intent to change it due to enabled authn/z in general.
As the kube-apiserver this sets the "system:unsecured" user info.

sttts commented:
/retest

sttts commented:
/retest

luxas commented on cmd/cloud-controller-manager/app/options/options.go:
```diff
@@ -89,23 +89,24 @@ func NewCloudControllerManagerOptions() (*CloudControllerManagerOptions, error)
 
+	s.Authentication.RemoteKubeConfigFileOptional = true
+	s.Authorization.RemoteKubeConfigFileOptional = true
+	s.Authorization.AlwaysAllowPaths = []string{"/healthz"}
```
:+1:

luxas commented on cmd/cloud-controller-manager/app/options/options.go:
```diff
@@ -61,9 +61,9 @@ type CloudControllerManagerOptions struct {
 
-	SecureServing *apiserveroptions.SecureServingOptions
+	SecureServing *apiserveroptions.SecureServingOptionsWithLoopback
 	// TODO: remove insecure serving mode
```
What's the timeframe for that with this implemented in v1.12? v1.14?

luxas reviewed (approved):
/lgtm
Thanks @sttts!
